### PR TITLE
CI: Add CI to test using gcc-12 on Intel Sapphire Rapids

### PR DESCRIPTION
**Pull request by r-devulap**
Add CI to test AVX512FP16 content on Intel SPR (ex: PR #23435 and #23351).

### CODE REVIEW

**Comment by r-devulap** [comment]
hmm, SDE SPR build taking forever.

**Comment by r-devulap** [comment]
Ah, I should build outside the SDE and only run the tests in it. Fixing it.

File: .github/workflows/build_test.yml [hunk 1]
```diff
@@ -408,3 +408,30 @@ jobs:
...
+    - name: Install dependencies
+      run: |
+        python -m pip install -r test_requirements.txt
+        sudo apt install gcc-12 g++-12
+    - name: Build
+      run: |
+        export CC=/usr/bin/gcc-12
+        export CXX=/usr/bin/g++-12
+        python setup.py install
+        sde -spr -- python runtests.py -n -v
```
Comment (mattip) [inline .github/workflows/build_test.yml:437]: Could you split this into two steps?

I think you may need to only run the ufunc and linalg tests, with a comment that this is very slow

Comment (r-devulap) [reply .github/workflows/build_test.yml]: Done.

**Comment by mattip** [comment]
Maybe try the equivalent `python runtests.py -n -- numpy/core/tests/test_umath.py`?

**Comment by mattip** [comment]
The tests now run in 22 minutes, which seems acceptable. I wouldn't expect tests to fail here, is there a problem with the emulation?

One style nit: I prefer `python -m pytest` instead of bare `pytest`, but feel free to ignore it if you want.

**Comment by r-devulap** [comment]
> The tests now run in 22 minutes, which seems acceptable. I wouldn't expect tests to fail here, is there a problem with the emulation?

It doesn't fail locally. I also don't see these errors: `AssertionError: FPU precision mode changed from 0x37f to 0x77f during the test`.  I am using SDE version 9.14, wonder if 9.7 has any issues. Lets try with latest version. 

> One style nit: I prefer `python -m pytest` instead of bare `pytest`, but feel free to ignore it if you want.

That's fine with me. I will fix it.

**Comment by r-devulap** [comment]
Looks like v9.14 fixed it.

**Comment by mattip** [comment]
Thanks @r-devulap

**Comment by seiko2plus** [comment]
Would it be better to combine it with `sde_simd_avx512_test` instead of having two actions?

**Comment by mattip** [comment]
I think we need two since we want to make sure both compilers (gcc12 and the default gcc in the manylinux image) pass tests

**Comment by seiko2plus** [comment]
Make sense, sorry for the noise. I just thought `sde_simd_avx512_test` uses the default compiler of `ubuntu-latest`.

**Comment by mattip** [comment]
>  I just thought `sde_simd_avx512_test` uses the default compiler of ubuntu-latest

Hmm. Maybe it does and we should be forcing it to use 9.3? In any case, the default gcc in ubuntu 22.04 is 11.3

**Pull request merged**